staging/kos/pkg/controllers/connectionagent/network_interface.go

```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package connectionagent
+
+import (
+	"fmt"
+	gonet "net"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	k8scache "k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+
+	netv1a1 "k8s.io/examples/staging/kos/pkg/apis/network/v1alpha1"
+	netfabric "k8s.io/examples/staging/kos/pkg/networkfabric"
+)
+
+var localIfcIDGenerator uint64
+
+// networkInterface is a convenience interface to simplify the high-level logic
+// of the connection agent by concealing the differences between local and
+// remote network interfaces (which implement it).
+type networkInterface interface {
+	canBeOwnedBy(*netv1a1.NetworkAttachment) bool
+	index() string
+	String() string
+}
+
+// localNetworkInterface wraps a network fabric LocalNetIfc and adds to it state
+// that the network fabric ignores but is relevant to the connection agent.
+type localNetworkInterface struct {
+	netfabric.LocalNetIfc
+	hostName       string
+	id             string
```

Comment:
This field is here only to avoid getting an ExecReport from a previous incarnation, right?  I prefer to not manage another set of IDs if we do not have to.  How about instead of `id string` we have `execReport *ExecReport`?  Let `launchCommand` be given `execReportReceiver **ExecReport` (from `&localIfc.execReport`) rather than `saveReport bool`.

Reply:
Yes, you described the scenario correctly.

> think of the PostCreateExec as something of secondary importance

I agree, I will update the comments on NetworkAttachmentSpec::PostCreateExec. 

> it looks like the localNetworkInterface's postDeleteExec is not updated in this scenario. I think it should be.

It is reasonable to assume that postDeleteExec cleans up what `postCreateExec` did, but because in this scenario NA1 and NA2 have different postCreateExecs updating the postDeleteExec would break that.

Reply:
OK, I can see the virtue in seeing PostDelete as paired with PostCreate.  OTOH, I am not sure I like the idea that a mistake can not be fixed.  Also, there is this difficult scenario: continue the scenario above with a restart of the connection agent then deletion of NA2.  The sync at startup after the restart will have only NA2's PostDeleteExec available.

Reply:
Also: since we are making the PostCreateExec have secondary importance, the NA's Status should include an echo of which PostCreateExec was run.

Reply:
> Also: since we are making the PostCreateExec have secondary importance, the NA's Status should include an echo of which PostCreateExec was run.

Yeah, makes sense.

About getting rid of IDs of localNetowrkInterface, you wrote:

>  I prefer to not manage another set of IDs if we do not have to. How about instead of id string we have execReport *ExecReport? Let launchCommand be given execReportReceiver **ExecReport (from &localIfc.execReport) rather than saveReport bool.

To elaborate a bit more, we still need some protection, because then localIfc.execReport gets accessed by two goroutines concurrently (the one processing the interface owner and the one launching the postCreateExec). It's a benign data race, but I don't like that anyway, I think it's not safe depending on what the compiler does. One solution I like is something along the lines of localIfc.execReport being a struct with the execReport which also embeds a mutex as opposed to a bare execReport.

Reply:
> OTOH, I am not sure I like the idea that a mistake can not be fixed. 

I don't understand what you're referring to, but if the mechanism to fix the mistake is to hope that the scenario we're discussing takes place (in the hypothesis that NA2's postDeleteExec is set for the recycled network interface) it's not a good mechanism, because whether that scenario takes place depends on things out of the user's control (e.g. timing of informer notifications and processing in the CA).

> Also, there is this difficult scenario: continue the scenario above with a restart of the connection agent then deletion of NA2. The sync at startup after the restart will have only NA2's PostDeleteExec available.

Using NA2's PostDeleteExec is hardly a solution to the problem. First, on the assumption that PostDeleteExecs clean up what a PostCreateExecs did NA2's PostDeleteExec does no good if it differs from NA1's. Second, even if what you're suggesting did work, the problem you outlined is a special (I would expect it to be extremely rare) case of a more general problem: If the connection agent restarts, the postDeleteExec of the NetworkInterfaces whose owners were deleted during the CA downtime are lost for good. If no new NetworkAttachment eligible to own the networkInterface exists, there's no PostDeleteExec to execute at all.

In general, I think that it would be ok to explicitly advertise to users that post/Create Execs are a best-effort mechanism not to rely upon. I agree with you though t

Reply:
> I think that is reasonable behavior, although I would say the current comments on NetworkAttachmentSpec::PostCreateExec do not suggest that. I like the idea that the PostCreateExec and its report are associated with the creation of the network interface in the fabric, and think of the PostCreateExec as something of secondary importance; it is not something that determines creation and deletion of network interfaces in the fabric but rather is a consequence of them. I would favor rewriting the comments on that field to clearly say that.

e2f6cb0 updates comments on `NetworkAttachmentSpec.postCreateExec` and `NetworkAttachmentSpec.postDeleteExec` by making it clear that they are not guaranteed to execute.

> Also: since we are making the PostCreateExec have secondary importance, the NA's Status should include an echo of which PostCreateExec was run.

8eaccb1 updates `NetworkAttachmentStatus.postCreateExecReports` by including a field that reports the command that executed.

> I prefer to not manage another set of IDs if we do not have to. How about instead of id string we have execReport *ExecReport? Let launchCommand be given execReportReceiver **ExecReport (from &localIfc.execReport) rather than saveReport bool.

b82b56d does that, but instead of passing a bare execReport passes a mutex-embedding struct to avoid data races.

> OK, I can see the virtue in seeing PostDelete as paired with PostCreate. OTOH, I am not sure I like the idea that a mistake can not be fixed. Also, there is this difficult scenario: continue the scenario above with a restart of the connection agent then deletion of NA2. The sync at startup after the restart will have only NA2's PostDeleteExec available.

Currently in that scenario the network interface is given NA2's PostDeleteExec:
https://github.com/MikeSpreitzer/kube-examples/blob/simplify-ca/staging/kos/pkg/controllers/connectionagent/connection_agent.go#L588 and 
https://github.com/MikeSpreitzer/kube-examples/blob/simplify-ca/staging/kos/pkg/controllers/connectionagent/network_interface.go#L105

Also, there's yet another problem with postCreateExec and postDeleteExec: they both run asynchronously (as they should, since they might take a while) wrt  network attachment processing from workers. Follows a scenario where this causes problems: NA1 is created and implemented as local by CA1, its postCreateExec begins execution; immediately after NA1 is deleted, CA1 deletes its network interface while postCreateExec is still running and postDeleteExec is launched and executes concurrently with the postCreateExec. They might interfere with each other. One solution is postponing deletion after the postCreateExec completes; this is easy to implement because at the end of postCreateExec's execution NA1's namespaced name is enqueued so the worker will realize that NA1 has been deleted. But I don't think the user would expect such behavior, if he/she deletes a networkattachment he/she expects it to get deleted no matter how much of the postCreateExec has executed, and as we agreed above postCreate/DeleteExecs are somewhat of secondary importance, they should not drive creation/deletion of network. OTOH, what I fear might happen is that the local state on the node is polluted by the interference of postCreate and postDelete Execs and this causes problems when processing future NetworkAttachments.